core/src/main/java/com/scottlogic/datahelix/generator/core/walker/decisionbased/RowSpecTreeSolver.java

```diff
@@ -51,45 +52,91 @@ public RowSpecTreeSolver(ConstraintReducer constraintReducer,
-            .map(Merged::get);
+            .filter(newNode -> !newNode.element().isContradictory())
+            .map(weighted -> new WeightedElement<>(weighted.element().get(), weighted.weight()));
 
         return flatMap(
             rootOnlyConstraintNodes,
-            this::reduceToRowNodes);
+            weightedConstraintNode -> reduceToRowNodes(
+                new WeightedElement<ConstraintNode>(
+                    weightedConstraintNode.element(),
+                    weightedConstraintNode.weight() / rootNode.weight())));
     }
 
-    private Merged<ConstraintNode> combineWithRootNode(ConstraintNode rootNode, ConstraintNode option) {
+    private WeightedElement<Merged<ConstraintNode>> combineWithRootNode(ConstraintNode rootNode, ConstraintNode option) {
         ConstraintNode constraintNode = rootNode.builder()
             .addDecisions(option.getDecisions())
             .addAtomicConstraints(option.getAtomicConstraints())
             .addRelations(option.getRelations())
             .build();
 
-        return treePruner.pruneConstraintNode(constraintNode, getFields(option));
+        double applicabilityOfThisOption = option.getAtomicConstraints().stream()
+            .mapToDouble(optionAtomicConstraint -> rootNode.getAtomicConstraints().stream()
+                .filter(rootAtomicConstraint -> rootAtomicConstraint.getField().equals(optionAtomicConstraint.getField()))
+                .filter(rootAtomicConstraint -> rootAtomicConstraint instanceof InSetConstraint)
+                .map(rootAtomicConstraint -> (InSetConstraint)rootAtomicConstraint)
+                .findFirst()
+                .map(matchingRootAtomicConstraint -> {
+                    double totalWeighting = matchingRootAtomicConstraint.legalValues.stream()
+                        .mapToDouble(InSetRecord::getWeightValueOrDefault).sum();
+
+                    double relevantWeighting = matchingRootAtomicConstraint.legalValues.stream()
+                        .filter(legalValue -> optionAtomicConstraint.toFieldSpec().canCombineWithLegalValue(legalValue.getElement()))
+                        .mapToDouble(InSetRecord::getWeightValueOrDefault).sum();
+
+                    return relevantWeighting / totalWeighting;
+                })
+                .orElse(1d))
+            .sum();
+
+        if (applicabilityOfThisOption > 1){
+            double applicabilityFraction = applicabilityOfThisOption - (int) applicabilityOfThisOption;
+            applicabilityOfThisOption = applicabilityFraction == 0
+                ? 1
+                : applicabilityFraction;
+        }
+
+        if (applicabilityOfThisOption == 0){
+            return new WeightedElement<>(
+                Merged.contradictory(),
+                1
+            );
+        }
+
+        return new WeightedElement<>(
+            treePruner.pruneConstraintNode(constraintNode, getFields(option)),
+            applicabilityOfThisOption
+        );
```

Comment:
This is very complicated.  Possibly needs refactored into a few well named methods.  Also as its complicated we might need a comment explaining what this thing is doing so someone who isn't familiar with this area can understand it.

Reply:
I've added some comments to the code and refactored some of the calculations into functions, the complexity is somewhat unavoidable i think.